### PR TITLE
ASLTS: Named References tests

### PR DESCRIPTION
**Pull request by Lekensteyn**
The spelling fixes should be non-controversial and can be cherry-picked already.

The new test needs some review, I looked at the `table` test and did some pattern matching to craft the new test, maybe I miss something. This test is the result of analysis in https://github.com/Bumblebee-Project/bbswitch/issues/142 (that should probably have happened on the ACPICA bugtracker though, but for now, see that link for AMLi logs and tested scenarios).

Ping @zetalog

Edit: reported the named ref issue at https://bugs.acpica.org/show_bug.cgi?id=1333

### CODE REVIEW

**Comment by Lekensteyn** [comment]
Test result shows failure as expected with the current implementation:
```
$ ../../generate/unix/bin/acpiexec -b "Execute MAIN" ./tmp/aml/20160930-64/opt/64/package.aml

Intel ACPI Component Architecture
AML Execution/Debug Utility version 20160930-64
Copyright (c) 2000 - 2016 Intel Corporation

Input file ./tmp/aml/20160930-64/opt/64/package.aml, Length 0x4550 (17744) bytes

ACPI: RSDP 0x00000000006872E0 000024 (v02 Intel )
ACPI: XSDT 0x0000000001F82FA0 000034 (v00 Intel  AcpiExec 00001001 INTL 20160930)
ACPI: FACP 0x0000000000687100 000114 (v05 Intel  AcpiExec 00001001 INTL 20160930)
ACPI: DSDT 0x0000000001F841E0 004550 (v02 Intel  Many     00000001 INTL 20160930)
ACPI: FACS 0x00000000006872A0 000040
Initializing General Purpose Events (GPEs):
    Initialized GPE 00 to 1F [_GPE] 4 regs on interrupt 0x0 (SCI)
    Initialized GPE 64 to 263 [_GPE] 64 regs on interrupt 0x0 (SCI)

Initializing Namespace objects:
Table [DSDT: Many    ] (id 01) -  397 Objects with   0 Devices,   0 Regions,   70 Methods (13/57/1 Serial/Non/Cvt)
ACPI: 1 ACPI AML tables successfully acquired and loaded

Completing Region/Field/Buffer/Package initialization:
    Initialized 0/0 Regions 0/0 Fields 0/0 Buffers 16/16 Packages (406 nodes)
Initializing Device/Processor/Thermal objects and executing _INI/_STA methods:
    Executed 0 _INI methods requiring 0 _STA executions (examined 2 objects)
Evaluating \MAIN
   tbfind-0223 [00] TbFindTable           : Found table [DSDT]
[ACPI Debug]  "64-bit mode"
[ACPI Debug]  "Release of parent ACPICA code 0x0000000020160930"
[ACPI Debug]  "Release of ASLTS test suite  0x0000000000000016"
[ACPI Debug]  "Settings of ASLTS test suite 0x0000000000000005"
[ACPI Debug]  "TEST (functional), Named references in PackageList"
[ACPI Debug]  "CHKR test started"
[ACPI Debug]  "---------- ERROR    : 0x00000000178B7000, 0x0000000000000000, CHKR"
[ACPI Debug]  "TITLE               : Named references in PackageList"
[ACPI Debug]  "COLLECTION          : functional"
[ACPI Debug]  "TEST CASE           : package"
[ACPI Debug]  "TEST                : CHKR"
[ACPI Debug]  "ERROR,    File      : namedref.asl"
[ACPI Debug]  "          Index     : 0x0000000000000000, (0)"
[ACPI Debug]  "**** Expected Result: 0x0000000000000002, (2)"
[ACPI Debug]  "**** Actual Result  : 0x000000000000000B, (11)"
[ACPI Debug]  "---------- END
"
[ACPI Debug]  "---------- ERROR    : 0x00000000178B7001, 0x0000000000000000, CHKR"
[ACPI Debug]  "TITLE               : Named references in PackageList"
[ACPI Debug]  "COLLECTION          : functional"
[ACPI Debug]  "TEST CASE           : package"
[ACPI Debug]  "TEST                : CHKR"
[ACPI Debug]  "ERROR,    File      : namedref.asl"
[ACPI Debug]  "          Index     : 0x0000000000000001, (1)"
[ACPI Debug]  "**** Expected Result: 0x0000000000000002, (2)"
[ACPI Debug]  "**** Actual Result  : 0x000000000000000B, (11)"
[ACPI Debug]  "---------- END
"
[ACPI Debug]  "---------- ERROR    : 0x00000000178B7002, 0x0000000000000000, CHKR"
[ACPI Debug]  "TITLE               : Named references in PackageList"
[ACPI Debug]  "COLLECTION          : functional"
[ACPI Debug]  "TEST CASE           : package"
[ACPI Debug]  "TEST                : CHKR"
[ACPI Debug]  "ERROR,    File      : namedref.asl"
[ACPI Debug]  "          Index     : 0x0000000000000002, (2)"
[ACPI Debug]  "**** Expected Result: 0x0000000000000002, (2)"
[ACPI Debug]  "**** Actual Result  : 0x0000000000000001, (1)"
[ACPI Debug]  "---------- END
"
[ACPI Debug]  ":CTST:functional:package:CHKR:FAIL:Errors # 03 00 00 00:"
[ACPI Debug]  "Run time (in seconds): 0x0000000000000000"
[ACPI Debug]  "The total number of exceptions handled: 0x0000000000000000"
[ACPI Debug]  "========= ERRORS SUMMARY (max 400):"
[ACPI Debug]  "package, namedref.asl, 0000000000000000"
[ACPI Debug]  "package, namedref.asl, 0000000000000001"
[ACPI Debug]  "package, namedref.asl, 0000000000000002"
[ACPI Debug]  "========= END."
[ACPI Debug]  "========= ROOT METHODS SUMMARY (max 600):"
[ACPI Debug]  ":STST:functional:package:CHKR:FAIL:Errors # 03 00 00 00:"
[ACPI Debug]  "========= END."
[ACPI Debug]  "TEST ACPICA: 64-bit : FAIL : Errors # 0x0000000000000003, Failed tests # 0x0000000000000001"
0x11 Outstanding allocations after evaluation of \MAIN
Evaluation of \MAIN returned object 0x1f7ea60, external buffer length 18
  [Integer] = 0000000000000001
```

For some reason `./aslts.sh` does not fail though, why is that?

@zetalog What is the story with `FULL/DECL.asl` and `functional/Makefile`, why is the `module` test not in that file?

**Comment by zetalog** [comment]
> What is the story with  FULL/DECL.asl  and  functional/Makefile , why is the  module  test not in that file?

It looks it was a problem caused by the fact that I never ran full tests.
I only ran specific aslts with -m -c options.
So "module" is currently not a part of the full run.

File: tests/aslts/src/runtime/collections/functional/package/RUN.asl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+if (STTT("Named references in PackageList", TCLF, 15, W01b)) {
```
Comment (zetalog) [inline tests/aslts/src/runtime/collections/functional/package/RUN.asl]: W01b is actually breaking the index of the test suite.
However, it seems it has already been broken.
So it's acceptable for now.

File: tests/aslts/HOW_TO_USE [hunk 2]
```diff
@@ -71,7 +71,7 @@ How to execute the run-time tests
    and the batch execute option. For example:
 
         > cd aslts/tmp/aml
-        > acpiexec -eMAIN 20090320/nopt/32/arithmetic.aml
+        > acpiexec -b "Execute MAIN" 20090320/nopt/32/arithmetic.aml
```
Comment (zetalog) [inline tests/aslts/HOW_TO_USE]: This looks a good fix to me.

File: tests/aslts/src/runtime/collections/functional/package/namedref.asl [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	Method(_ON) {}
+	Method(_OFF) {}
+}
+Name(INT1, 1)
+
+Name(PKG1, Package() {
+	PWR1,
+	PWR2,
+	INT1
+})
```
Comment (zetalog) [inline tests/aslts/src/runtime/collections/functional/package/namedref.asl]: I'm trying to figure out what's going on here:

ComputationalData := ByteConst | WordConst | DWordConst | QWordConst | String |
ConstObj | RevisionOp | DefBuffer
DataObject := ComputationalData | DefPackage | DefVarPackage
DataRefObject := DataObject | ObjectReference | DDBHandle

DefPackage := PackageOp PkgLength NumElements PackageElementList
PackageOp := 0x12
DefVarPackage := VarPackageOp PkgLength VarNumElements PackageElementList
VarPackageOp := 0x13
NumElements := ByteData
VarNumElements := TermArg => Integer
PackageElementList := Nothing | <PackageElement PackageElementList>
PackageElement := DataRefObject | NameString

It looks the PackageElement really can have "ObjectReference".
I guess we need to check all other DataRefObject cases.

**Comment by Lekensteyn** [comment]
> I'm trying to figure out what's going on here:
> It looks the PackageElement really can have "ObjectReference".

Let's start with the Package definition in the **AML grammar** and gradually expand every token (according to the ACPI 6.1 spec):
```
PackageElementList := Nothing | <PackageElement PackageElementList>
PackageElement := DataRefObject | NameString

// Expand PackageElement
DataRefObject := DataObject | ObjectReference | DDBHandle
NameString :=  <RootChar NamePath> | <PrefixPath NamePath>

// Expand NameString (nothing unusual, basically encodes \NAME, ^^^SOME.NAME, etc.
RootChar := '\'
NamePath := NameSeg | DualNamePath | MultiNamePath | NullName
PrefixPath := Nothing | <'^' PrefixPath>

// Expand DataRefObject
DataObject := ComputationalData | DefPackage | DefVarPackage
ObjectReference := /* not defined in AML grammar, Integer in ASL */
DDBHandle := /* not defined in AML grammar, Integer in ASL */

// Expand DataObject
ComputationalData := ByteConst | WordConst | DWordConst | QWordConst | String | ConstObj | RevisionOp | DefBuffer
DefPackage := PackageOp PkgLength NumElements PackageElementList
PackageOp := 0x12
DefVarPackage := VarPackageOp PkgLength VarNumElements PackageElementList
VarPackageOp := 0x13
```

Indeed, in **AML** the PackageElement can be ObjectReference, but `ObjectReference` is only defined in the **ASL** grammar (is this an error?). Now consider these two examples:

    Name(_PR0, Package() { PWR1 })
    Name(_PR2, Package() { "PWR1" })

In the first case, `PWR1` is matched by the NameString definition.
In the second case, `"PWR1"` is matched by `DataRefObject => DataObject => ComputationalData => String`.
In both cases, ObjectReference is irrelevant (not matching!).

Let's put AML aside and look at the **ASL interpretation**.
```
PackageTerm := Package ( NumElements ) { PackageList } => Package
PackageList := Nothing | <PackageElement PackageListTail>
PackageListTail := Nothing | <CommaChar PackageElement PackageListTail>
PackageElement := DataObject | NameString // !!! DataObject, *not* DataRefObject

// Expand PackageElement, DataObject
DataObject := BufferData | PackageData | IntegerData | StringData
BufferData := Type5Opcode | BufferTerm
PackageData := PackageTerm
IntegerData := Type3Opcode | Integer | ConstTerm
StringData := Type4Opcode | String
```
According to ASL grammar above, a `PackageElement` *cannot* be a `DataRefObject (=> ObjectReference)`, only `DataObject` is valid (besides `NameString`). Another error in the specification: in Section 19.6.82 Method says (emphasis mine):
> MethodName is evaluated as a **Namestring data type**.

Table 19-367 however does not define such a "data type". I guess that lacking a "Namestring" type, Microsoft decided to implicitly convert a `NameString` to a `String`.

ACPICA should not try to dereference at table load time, but store a string instead for NameStrings in Package elements. How can I ask for a change/clarification in the ACPI spec?

**Comment by Lekensteyn** [comment]
@acpibob @zetalog
I have added a flag to test for the bug (bdemo) but no steps are reported as pass/fail (even if I enable the flag). Why is that?

Command was `./asltsh.sh -c bdemo -u`, part of Summary logfile:
```
64-bit norm mode:

bdemo:
                   (16:13:20-16:13:27 (00:07.00))
       PASS    :  0
       FAIL    :  0
       BLOCKED :  0
       SKIPPED :  0
         total :   0
         Outstanding allocations after execution : 0
         Outstanding allocations (ACPI Error)    : 0
         Memory statistics (per cache type):
           Total            : 0 Kb
           Maximum occupied : 0(Kb) 0 0 0 0 0
           Outstandings     : 0 0 0 0 0 0
TOTAL:             (64-bit norm mode, 2016-11-15 16:13:20, 2016-11-15 16:13:27 (00:07.00))
    PASS       :  0
    FAIL       :  0
    BLOCKED    :  0
    SKIPPED    :  0
    Tests      :   0
    Test Cases       : 1 (of 47)
    Test Collections : 1 (of 8),   bdemo
    Outstanding allocations after execution : 0
    Outstanding allocations (ACPI Error)    : 0
    Large Reference Count   (ACPI Error)    : 0
    Memory consumption total                : 0 Kb
```

**Comment by acpibob** [comment]
Just so that we are on the same page with this, please see ACPI 6.1 specification, section 19.6.99. A serious attempt was made to clarify this entire issue.

**Comment by acpibob** [comment]
>> I have added a flag to test for the bug (bdemo) but no steps are reported as pass/fail (even if I enable the flag). Why is that?

There are a number of issues like this in the suite at this time. Some things have been broken. We have someone working on these right now.

**Comment by acpibob** [comment]
Please split out the spelling fixes into another commit, and we can take it immediately.
thanks,
bob

**Comment by Lekensteyn** [comment]
> Just so that we are on the same page with this, please see ACPI 6.1 specification, section 19.6.99. A serious attempt was made to clarify this entire issue.

Yes, that is the same page. Depending on which parts you read, it seems ambiguous. Based on these parts, an object reference seems allowed.

> PackageList contains the data items, constants, and/or **object references** used to initialize the package

> There are three types of package elements allowed in the PackageList: ... and **named references to objects** other than Data Objects.

Based on these parts, it is not so clear that object references are allowed (it could as well be a string):

> These Named References to non-Data Objects cannot be resolved to values. They are instead returned in the package as references:

and from 19.2.3 (ASL grammar):
> PackageElement := DataObject | NameString

Thanks for looking into it.

**Comment by acpibob** [comment]
>>>PackageElement := DataObject | NameString

The NameString is in fact a reference. An ASCII String (with quotes) is a DataObject.

Package () { 1, 2, \_SB_.PCI0, 3}
In this case, the third element is a NameString that is a reference to \_SB_. PCI0

Package () {1, 2, "\_SB_.PCI0", 3}
In this case, the third element is an ASCII string that is simply a DataObject

**Comment by zetalog** [comment]
> According to ASL grammar above, a  PackageElement  cannot be a  DataRefObject (=> ObjectReference) , only  DataObject  is valid (besides  NameString ). 

However, for NameString, there are many cases in AML grammar that they'll be resolved to the namespace object (though IMO, such resolution should be done at per-opcocde executer level) while in ACPICA this is in the generic code path...
Hmm..., OTOH, ACPICA really doesn't know if the grammar primitive is a NameString or something else. We are lacking of such information in AcpiGbl_AmlOpInfo, that's why I entirely re-implement this in my from-scratch interpreter here:
http://sdfirm.sourceforge.net
Where I put lots of effort re-implementing basic AML primitives.
Well, that makes the problem much more difficult to solve than anyone can expect...

**Comment by zetalog** [comment]
Since something has been merged, could you rebase and rename this pull request?

**Comment by Lekensteyn** [comment]
I've rebased the branch, but it is not suitable for merging. It just shows possible ways to implement the tests (note that the demob tests are broken, failures do not seem to have any effect on the final result).
(Note that there is one commit, a revert thereof an a smaller demob change that is broken per above.)

**Comment by acpibob** [comment]
as far as the question about 0274 and aslts.sh, I believe that the entire directory bdemo/ACPICA is not executed as part of the normal test suite, they are examples of known issues.

**Comment by Lekensteyn** [comment]
Okay, that has just bitrotten badly.

@zetalog Are you working on tests as part of https://bugs.acpica.org/show_bug.cgi?id=1333? If so, then I think that this PR can be closed

**Comment by acpibob** [comment]
These issues are still under investigation and analysis, so nothing should be closed at this time.

**Pull request closed without merging**